services/src/main/scala/cromwell/services/metadata/MetadataService.scala

```diff
@@ -198,39 +198,55 @@ object MetadataService {
 
   implicit private class EnhancedWomTraversable(val womValues: Iterable[WomValue]) extends AnyVal {
-    def toEvents(metadataKey: MetadataKey): List[MetadataEvent] = if (womValues.isEmpty) {
+    def toEvents(metadataKey: MetadataKey, fileMap: Map[Path, Path]): List[MetadataEvent] = if (womValues.isEmpty) {
       List(MetadataEvent.empty(metadataKey.copy(key = s"${metadataKey.key}[]")))
     } else {
       womValues.toList.zipWithIndex
         .flatMap { case (value, index) =>
-          womValueToMetadataEvents(metadataKey.copy(key = s"${metadataKey.key}[$index]"), value)
+          womValueToMetadataEvents(metadataKey.copy(key = s"${metadataKey.key}[$index]"), value, fileMap)
         }
     }
   }
 
-  def womValueToMetadataEvents(metadataKey: MetadataKey, womValue: WomValue): Iterable[MetadataEvent] = womValue match {
-    case WomArray(_, valueSeq) => valueSeq.toEvents(metadataKey)
+  def womValueToMetadataEvents(metadataKey: MetadataKey,
+                               womValue: WomValue,
+                               fileMap: Map[Path, Path] = Map.empty
+  ): Iterable[MetadataEvent] = womValue match {
+    case WomArray(_, valueSeq) => valueSeq.toEvents(metadataKey, fileMap)
     case WomMap(_, valueMap) =>
       if (valueMap.isEmpty) {
         List(MetadataEvent.empty(metadataKey))
       } else {
         valueMap.toList flatMap { case (key, value) =>
-          womValueToMetadataEvents(metadataKey.copy(key = metadataKey.key + s":${key.valueString}"), value)
+          womValueToMetadataEvents(metadataKey.copy(key = metadataKey.key + s":${key.valueString}"), value, fileMap)
         }
       }
     case objectLike: WomObjectLike =>
       if (objectLike.values.isEmpty) {
         List(MetadataEvent.empty(metadataKey))
       } else {
         objectLike.values.toList flatMap { case (key, value) =>
-          womValueToMetadataEvents(metadataKey.copy(key = metadataKey.key + s":$key"), value)
+          womValueToMetadataEvents(metadataKey.copy(key = metadataKey.key + s":$key"), value, fileMap)
         }
       }
     case WomOptionalValue(_, Some(value)) =>
-      womValueToMetadataEvents(metadataKey, value)
+      womValueToMetadataEvents(metadataKey, value, fileMap)
     case WomPair(left, right) =>
-      womValueToMetadataEvents(metadataKey.copy(key = metadataKey.key + ":left"), left) ++
-        womValueToMetadataEvents(metadataKey.copy(key = metadataKey.key + ":right"), right)
+      womValueToMetadataEvents(metadataKey.copy(key = metadataKey.key + ":left"), left, fileMap) ++
+        womValueToMetadataEvents(metadataKey.copy(key = metadataKey.key + ":right"), right, fileMap)
+    case file: WomSingleFile =>
+      // Our lookup key is a string; to avoid exceptions, stringify paths instead of pathifying the string
+      val stringifiedMap: Map[String, String] = fileMap map { case (src: Path, dst: Path) =>
+        src.pathAsString -> dst.pathAsString
+      }
```

Comment:
Nitpick you can feel free to ignore: we're likely stringifying this map over and over, right? Is the savings we get from passing a pre-stringified map around greater than the grossness of doing that?

Reply:
I thought about this and wanted to avoid [stringly typed](https://wiki.c2.com/?StringlyTyped) as much as possible. Looking at the `pathAsString` code for GCS, there will definitely be some string manipulation and concatenation, but nothing computationally severe like regex.

https://github.com/broadinstitute/cromwell/blob/9b98728520850b195f28cb02c652a8df24838f3e/filesystems/gcs/src/main/scala/cromwell/filesystems/gcs/GcsPathBuilder.scala#L208-L212